### ISSUE TITLE
Train and publish Qwen-Mathia v1 from the agnostic corpus

### ISSUE THREAD

**Issue by murillo128** [murillo128/mathia#47]
Research umbrella: #2
Conceptual-specialist line: #29
Agnostic corpus: #44
Riemann corpus complement: #42 — **out of scope for this run**
Model identity: `docs/MODEL_AND_COMPUTE_CONSTRAINTS.md`

# Goal

Train the first **domain-agnostic Qwen-Mathia** checkpoint from the accepted `agnostic-mathia-full-v1` corpus and publish the resulting reproducible PEFT model under the `murillo2000` Hugging Face namespace.

This is the first bounded Mathia post-training run. Its product is a concrete local conceptual specialist that can be validated later against qwen-lean and can later serve as the parent/control for a separate Riemann-specialization experiment.

The training target is deliberately narrow:

```text
Qwen/Qwen3-8B-Base
        |
        |  agnostic Mathia SFT
        v
Qwen-Mathia-v1
        |
        `--> publish frozen PEFT adapter + provenance on Hugging Face
```

**Do not add the Riemann-Mathia corpus in this issue. Do not run the qwen-lean fertility validation here.** The purpose is to produce and publish the first agnostic Mathia checkpoint cleanly; scientific validation is a later experiment.

## Explicit sequencing override

Older #29/#32 text says not to start Mathia training before the intuition-fertility channel is validated. The user has now explicitly chosen a different sequencing on **2026-08-20**: train Qwen-Mathia first and defer the qwen-lean validation until afterward.

For this bounded issue, that user decision supersedes the old compute gate. It does **not** invalidate the scientific concerns preserved by #29/#32 and does not make teacher similarity, training loss, or fluent Mathia prose a success metric.

# Frozen model ancestor

Use exactly the common ancestor already frozen by Mathia and qwen-lean:

```text
model_id:           Qwen/Qwen3-8B-Base
model_revision:     49e3418fbbbca6ecbdf9608b4d22e5a407081db4
tokenizer_id:       Qwen/Qwen3-8B-Base
tokenizer_revision: 49e3418fbbbca6ecbdf9608b4d22e5a407081db4
```

No instruct/chat checkpoint substitution. No silent model/tokenizer revision drift. No chat template in the primary lane.

The upstream model reports Apache-2.0 licensing. Record the exact upstream model-card/license evidence used for publication.

# Frozen corpus input

Consume only the accepted domain-agnostic release on `main`:

```text
release_id:  agnostic-mathia-full-v1
freeze_id:   freeze_eeeeb89af3d2ac75d1ff5dad5623b63d1d24dfbddb965beca2f1c4aac9f9867f
contract:    mathia-interchange-v1
```

Important frozen hashes from `experiments/agnostic_mathia_corpus/release_v1/freeze.json`:

```text
records.jsonl
  d04e090126f58f039b229053cb456967cb9b6d209fdb172c9969a22909e483c1

rendered_trainable.jsonl
  9cbc2c87637a4139234acb6de85b3f863542503512319bf21bf11b27f223f625

trainable_manifest.json
  c5568184d3e55e753a8f271d6d2d2bb55fe0eb7e5b21373bc8ff044001d1ff6c

review_content_freeze.json
  2fa415dd4e0dc94a7c7f63ec7742c648415d1e6be41dc04b209603029adb1795
```

The release currently reports 214 trainable objects overall. Recompute all role/count/token statistics from the frozen manifest before training rather than trusting hard-coded counts.

Do not regenerate teacher interpretations, alter quality decisions, repair source material, or add new corpus records in this issue.

# What to supervise

The primary hypothesis for this first run is that the useful training signal is the **conceptual transformation**, not ordinary language-model continuation of mathematical source text.

## Primary optimizer examples

Use only accepted/eligible:

- `interpretation` objects: mathematical material + common conceptual-reading task -> Mathia response;
- `synthesis` objects: multiple mathematical materials + synthesis task -> Mathia response.

`source` objects remain the mathematical input/provenance from which those examples are rendered but are **not standalone optimizer targets in the primary v1 lane**. This avoids weighting the same source material once as an LM continuation and again inside its interpretation.

Expected from the current release is approximately 98 accepted interpretations + 18 accepted syntheses; verify the exact counts programmatically from the frozen objects/manifest and record them in the training evidence.

## Loss masking

Use completion-only causal-LM loss:

```text
mathematical material + task   -> masked from loss
Mathia response                -> supervised
EOS                            -> supervised
```

For synthesis, all source materials and the task are prompt/context and only the synthesis response is supervised.

Reuse the canonical `mathia-interchange-v1` renderer rather than inventing a second visible corpus dialect. Implement only the minimal deterministic prompt/response boundary needed to mask labels.

No data augmentation, teacher rewriting, paraphrase multiplication, synthetic prompt variants, or Riemann examples in this first run.

# Checkpoint A — materialize and audit the exact training workload

Before GPU training:

1. verify all frozen corpus hashes above;
2. resolve every optimizer example through `mathia-interchange-v1`;
3. prove that every selected object is `accepted` + training-eligible;
4. prove that no `rejected`, `quarantined`, or `evaluation_only` object enters optimizer batches;
5. record exact object IDs and deterministic ordering/shuffle seed;
6. tokenize with the exact frozen Qwen tokenizer and report prompt/response/total token distributions separately for interpretation and synthesis;
7. determine the true maximum sequence length;
8. **never silently truncate** an accepted example;
9. materialize a compact immutable training-manifest artifact with IDs, hashes, role, token lengths, and loss-mask boundaries;
10. add tests that changing private provenance/quality metadata does not change model-visible bytes, preserving the existing interchange guarantee.

No train/validation split is required for scientific model selection in this issue. The corpus is small and later downstream validation is intentionally separate. Train on the complete selected conceptual workload and preserve multiple checkpoints instead of choosing a checkpoint by an ad-hoc local benchmark.

# Checkpoint B — GPU preflight and frozen training recipe

Target hardware is the existing project **NVIDIA RTX 4000 Ada Generation 20 GB** unless execution occurs on an explicitly documented equivalent/better local GPU.

Use the successful qwen-lean Qwen3-8B QLoRA recipe as the engineering prior, but do not import any Lean data or semantics.

Primary lane:

```text
base load:            4-bit NF4
compute dtype:        bfloat16
double quantization:  true
PEFT:                  LoRA
LoRA rank:             16
LoRA alpha:            32
LoRA dropout:          0.0
bias:                  none
target modules:        q_proj, k_proj, v_proj, o_proj,
                       gate_proj, up_proj, down_proj
gradient checkpointing:true
optimizer:             paged_adamw_8bit
packing:               false
truncation:            false
seed:                  0
```

Use a conservative first-run learning schedule because the conceptual corpus is tiny compared with the qwen-lean SFT corpus:

```text
learning_rate:                 5e-5
weight_decay:                  0.0
max_grad_norm:                 1.0
lr_schedule:                   cosine
per_device_micro_batch_size:   1
gradient_accumulation_steps:   8
epochs:                        4
```

Choose `maximum_sequence_tokens` only after the exact tokenizer audit in Checkpoint A. Use the smallest clean bound that contains **every** selected example with no truncation. If any selected example cannot fit a technically reasonable context on the 20 GB Ada under the declared QLoRA lane, stop and surface the concrete blocker rather than silently dropping it.

Run a bounded smoke first using both interpretation and synthesis examples. The smoke must demonstrate:

- model/tokenizer exact revision;
- finite forward/backward loss;
- non-zero adapter gradients;
- no prompt-token supervision leakage;
- no sequence truncation;
- adapter save/reload works;
- deterministic manifest/order with seed 0;
- measured peak GPU memory leaves a safe execution margin.

The smoke is operational only. Do not inspect its generations and tune the learning rate/epochs to make the prose look better.

If the declared QLoRA lane itself fails for an implementation/runtime reason, fix the bounded implementation problem and rerun. Do not substitute another parent model.

# Checkpoint C — full Qwen-Mathia-v1 training

After the preflight passes, execute exactly one frozen full run over the complete selected workload.

Preserve checkpoints at least after cumulative epochs:

```text
epoch 1
epoch 2
epoch 4  <- terminal/root published adapter
```

Do not early-stop or change hyperparameters based on subjective generation quality. Training loss is operational evidence only, not evidence that Mathia capability improved.

Record at minimum:

- exact repository commit and dirty-worktree status;
- model/tokenizer IDs and immutable revisions;
- corpus release/freeze IDs and all material hashes;
- exact selected object IDs/counts by role;
- tokenizer/token statistics;
- package versions (`torch`, `transformers`, `trl`, `peft`, `bitsandbytes`, etc.);
- CUDA/driver/GPU identity;
- training config hash;
- optimizer-step count and examples/tokens seen;
- loss curve by optimizer step/epoch;
- learning-rate curve;
- wall time;
- measured peak GPU memory;
- adapter parameter count;
- every saved checkpoint path and SHA-256/hash inventory;
- any warnings/retries/interruptions.

An interrupted run may resume only from an exact saved state with provenance showing that the resumed result is semantically the same frozen run. Do not silently restart with changed ordering/config.

# Checkpoint D — post-training technical sanity only

This issue does **not** run the scientific Mathia/qwen-lean validation.

After training, do only bounded technical checks:

- load the terminal adapter from disk on the exact base;
- deterministic greedy generation on a tiny frozen smoke set to prove inference plumbing works;
- confirm outputs are non-empty and the model terminates under the declared token cap;
- prove base-only and adapter-enabled execution are distinguishable at the weight/config level;
- verify adapter/config hashes after reload.

Do not promote these outputs into a quality score, compare them to Codex, or select a different checkpoint because it sounds better.

Keep epoch-1 and epoch-2 checkpoints so later validation can measure training-depth effects without rerunning training.

# Checkpoint E — freeze publication artifact

The primary published artifact is a **PEFT adapter**, not an unnecessary merged 8B full-weight duplicate.

Freeze the terminal epoch-4 adapter and its complete publication bundle before upload. Include at minimum:

```text
adapter_model.safetensors
adapter_config.json
README.md / model card
training_config.*
training_manifest.*
training_summary.*
checkpoint hash inventory
```

The adapter config must bind the exact upstream base model/revision. Do not copy or modify the upstream tokenizer unless a concrete technical requirement exists; this run should not change tokenizer vocabulary or special tokens.

Intermediate epoch-1 and epoch-2 adapters should remain recoverable. Prefer publishing them under deterministic `checkpoints/epoch-1` and `checkpoints/epoch-2` paths in the same model repository, or another equally simple immutable layout. The root adapter is the terminal epoch-4 result.

A merged full-weight checkpoint is optional and **not required** for acceptance. Do not publish one merely for convenience.

# Checkpoint F — publish on Hugging Face

Authenticated Hugging Face identity observed during issue design:

```text
namespace: murillo2000
```

Target model repository:

```text
murillo2000/qwen3-8b-base-mathia-v1
```

Before upload, the executor must run `hf auth whoami` in the execution environment and confirm that the token has write permission to this namespace. The current ChatGPT/Hugging Face connection establishes account identity but must not be assumed to provide the executor's upload credential.

Use the current `hf` CLI, not deprecated `huggingface-cli` commands.

Recommended publication flow after the exact artifact is reviewed:

```text
hf repos create murillo2000/qwen3-8b-base-mathia-v1 --exist-ok
hf upload murillo2000/qwen3-8b-base-mathia-v1 <publication-dir> \
  --commit-message "Publish Qwen-Mathia v1 adapter"
```

Use `hf upload-large-folder` only if the actual artifact layout makes resumable large-folder upload preferable.

## Model card requirements

The Hub model card must state clearly:

- this is **Qwen-Mathia v1**, a research PEFT/QLoRA conceptual-mathematics adapter;
- exact base: `Qwen/Qwen3-8B-Base@49e3418f...`;
- exact Mathia Git commit/PR and corpus freeze IDs/hashes;
- training used only `agnostic-mathia-full-v1`;
- **no Riemann-Mathia data** was used;
- role counts and token counts actually supervised;
- response-only loss semantics;
- exact LoRA/quantization/training config;
- hardware/runtime and wall time;
- how to load with `transformers` + `peft`;
- intermediate checkpoints available and root adapter identity;
- training loss/technical sanity results only;
- **no claim yet that the model improves qwen-lean, theorem proving, conceptual transfer, or mathematical reasoning**;
- later qwen-lean validation is separate work;
- known risk of teacher-style distillation / small-corpus overfitting;
- provenance/licensing boundaries of the corpus and upstream model.

The upstream base is Apache-2.0. Before setting a Hub `license:` field for the adapter, perform an explicit publication/provenance check against the corpus release's licensing boundaries. Do not infer a global dataset license that the corpus itself does not grant. If Apache-2.0 remains appropriate for the adapter artifact after that audit, record the rationale; otherwise use accurate Hub metadata rather than guessing.

Do **not** upload restricted/raw source artifacts or external corpus stores to the model repository.

Suggested tags where accurate:

```text
peft
safetensors
qwen3
lora
qlora
mathematics
reasoning
text-generation
base_model:Qwen/Qwen3-8B-Base
base_model:adapter:Qwen/Qwen3-8B-Base
```

# Checkpoint G — verify the published model from an immutable Hub revision

Publication is not complete merely because upload returned success.

After upload:

1. record the exact Hugging Face commit/revision;
2. clear/use a clean temporary cache;
3. download the adapter from that immutable Hub revision;
4. verify published file hashes against the frozen local publication bundle;
5. load `Qwen/Qwen3-8B-Base@49e3418f...` + the published adapter with no local-path fallback;
6. run the same deterministic technical smoke prompt(s);
7. confirm the Hub model card renders and contains exact provenance;
8. record the public model URL + immutable revision in repository evidence.

No mutable `main`-only reference is sufficient for final reproducibility; preserve the exact HF commit hash.

# GitHub delivery

Deliver one focused PR linked to this issue containing the reusable training/publishing support and compact evidence needed to reproduce the run, for example:

```text
experiments/qwen_mathia_v1/
  config.*
  training_manifest.*
  train.*
  publish.*
  evidence/
    preflight.*
    training_summary.*
    artifact_hashes.*
    huggingface_publication.*
```

Exact paths may differ if an equally simple existing repository pattern is better. Do not introduce a general ML platform or permanent Mathia training DSL for one run.

Keep bulky model weights/caches/checkpoints out of Git; Git retains configs, hashes, summaries and the exact Hub revision.

Normal repository tests should remain CPU-only. GPU training is an explicit execution step, not something CI should accidentally rerun.

# Independent review

Before public Hugging Face publication, run a fresh read-only review over the exact frozen training/publication target. Check at least:

- exact base/tokenizer identity;
- corpus freeze/hash integrity;
- selected object roles and no excluded/evaluation leakage;
- response-only loss mask correctness;
- no silent truncation;
- QLoRA/LoRA config and actual adapter targets;
- reproducible optimizer steps/order/seeds;
- checkpoint/artifact hashes;
- no Riemann data;
- licensing/publication boundaries;
- model-card claims do not outrun evidence.

After Hub upload, perform a second compact integrity check of the exact immutable HF revision and local-vs-Hub hashes. This second check is publication integrity, not a new model-quality review.

# Out of scope

- qwen-lean fertility/proof-uplift validation;
- benchmark-driven checkpoint selection;
- Riemann-Mathia training or mixing;
- qwen-lean-Riemann;
- RL/GRPO/verifier reward;
- weight merging with qwen-lean;
- new corpus generation or teacher calls;
- concept taxonomy redesign;
- full-model fine-tuning;
- merged full-weight publication unless independently justified;
- claims of mathematical improvement from training loss or prose quality.

# Exit decision

Record exactly one final state:

- `QWEN_MATHIA_V1_PUBLISHED` — frozen training completed, PEFT artifact verified, public Hugging Face immutable revision verified, and repository evidence is ready;
- `QWEN_MATHIA_TRAINING_BLOCKER` — the frozen training run cannot be completed credibly;
- `QWEN_MATHIA_DATA_BLOCKER` — frozen corpus/rendering/loss-boundary integrity prevents training;
- `QWEN_MATHIA_HARDWARE_BLOCKER` — the exact 8B training lane cannot run safely on available hardware;
- `QWEN_MATHIA_PUBLICATION_BLOCKER` — a valid local adapter exists but Hugging Face authentication/licensing/integrity prevents publication.

# Exit gate

Close only when either:

1. `QWEN_MATHIA_V1_PUBLISHED` is recorded and all of the following exist: exact training config + manifest, terminal and intermediate adapter hashes, compact training evidence, fresh independent review, public `murillo2000/qwen3-8b-base-mathia-v1` immutable HF revision, clean-download/reload/hash verification, and accurate model card; or
2. one blocker state above is supported by concrete evidence without silently changing parent model, corpus, or scientific scope.

Do not merge automatically from the executor workflow.

**Comment by murillo128** [comment]
## Completion summary

Issue #47 completed through PR #48, squash-merged to `main` as `8876cccb036f79ea6beba9b8ae692c4713d2c835`.

- Parent/base: `Qwen/Qwen3-8B-Base@49e3418fbbbca6ecbdf9608b4d22e5a407081db4`.
- Training corpus: frozen `agnostic-mathia-full-v1` only; no Riemann-Mathia records.
- Optimizer workload: 98 accepted interpretations + 18 accepted syntheses; source objects remained context/provenance only.
- Objective: response/EOS-only causal-LM loss; prompt masked; no packing or truncation; true maximum 727 tokens under the frozen 768-token bound.
- Training: NF4 QLoRA, LoRA r=16/alpha=32 over attention + MLP projections, 4 epochs, 60 optimizer steps, 464 example presentations, 135,944 total token presentations and 52,868 supervised response/EOS tokens.
- Hardware: NVIDIA RTX 4000 Ada Generation 20 GB; preflight passed with finite loss/gradients, confined adapter updates, save/reload, and safe memory margin.
- Published artifact: `murillo2000/qwen3-8b-base-mathia-v1` as a PEFT adapter. Immutable Hub revision: `a0dade8a2ce5222b1222f762434ca89528587f69`.
- Publication verification: local/Hub hashes match, immutable Hub adapter reload passes, deterministic technical generation hashes match, and required provenance/model-card markers are present.
- Licensing metadata is conservatively `license: other` because the corpus retains mixed source-specific licensing boundaries.
- Scientific boundary: no qwen-lean fertility, theorem-proving, transfer, or mathematical-reasoning claim was tested here. Those remain separate validation work.

Final technical state: `QWEN_MATHIA_V1_PUBLISHED`.
